hyper-capi/include/hyper.h

```diff
@@ -0,0 +1,318 @@
+
+// Set the poll function for this body.
+//
+// This function will be called each time more data is desired to write to
+// the transport. Use `hyper_body_set_data` to set the `userdata` argument.
+void hyper_body_set_poll(hyper_body *body,
+                         hyper_poll (*func)(void *userdata,
+                                      hyper_context *ctx));
+
+// Return a task that will yield the next chunk of bytes of the body, when
+// available.
+//
+// This task has a return type tag of `HYPER_TASK_BODY_NEXT`, which will
+// give a value of `hyper_buf *`.
+//
+// When the body is complete, the task's value will be `NULL`.
+hyper_task *hyper_body_next(hyper_body *body);
+
+// Return a task that will poll the body and execute the callback with each
+// body chunk that is received.
+//
+// The `hyper_buf` pointer is only a borrowed reference, it cannot live outside
+// the execution of the callback. You must make a copy to retain it.
+hyper_task *hyper_body_foreach(hyper_body *body,
+                               hyper_iter_step (*func)(void *userdata,
+                                                       const hyper_buf *buf),
+                               void *userdata);
+
+// TODO: hyper_body_trailers()
+
+
+// Get a pointer to the bytes in this buffer.
+//
+// This should be used in conjunction with `hyper_buf_len` to get the length
+// of the bytes data.
+//
+// This pointer is borrowed data, and not valid once the `hyper_buf` is
+// consumed/freed.
+const uint8_t *hyper_buf_bytes(const hyper_buf *buf);
+
+// Get the length of the bytes this buffer contains.
+size_t hyper_buf_len(const hyper_buf *buf);
+
+// Free this buffer.
+void hyper_buf_free(hyper_buf *buf);
+
+// Futures and Executors
+
+// Creates a new task executor.
+hyper_executor *hyper_executor_new(void);
+
+// Push a task onto the executor.
+hyper_error hyper_executor_push(hyper_executor *executor, hyper_task *task);
+
+// Polls the executor, trying to make progress on any tasks that have notified
+// that they are ready again.
+//
+// If ready, returns a task from the executor that has completed.
+//
+// If there are no ready tasks, this returns `NULL`.
+hyper_task *hyper_executor_poll(hyper_executor *executor);
+
+// Frees an executor, and any tasks it may currently be holding.
+void hyper_executor_free(hyper_executor *executor);
+
+typedef enum {
+	HYPER_TASK_EMPTY,
+	HYPER_TASK_ERROR,
+	HYPER_TASK_CLIENTCONN,
+	HYPER_TASK_RESPONSE,
+} hyper_task_return_type;
+
+// Query the return type of this task.
+hyper_task_return_type hyper_task_type(hyper_task *task);
+
+// Set a user data pointer to be associated with this task.
+//
+// This value will be passed to task callbacks, and can be checked later
+// with `hyper_task_userdata`.
+void hyper_task_set_data(hyper_task *task, void *userdata);
+
+// Retrieve the userdata that has been set via `hyper_task_set_data`.
+void *hyper_task_userdata(hyper_task *task);
```

Comment:
Wouldn't it be more consistent if it was `hyper_task_set_userdata` as well? I mean then "userdata" would be common among the function names and the argument names.

Reply:
That's true. In the rest of the C API, it's been just `*_set_data`, but then wondered if `hyper_task_data` would be to similar to `hyper_task_value`.